### PR TITLE
Disable test values in BNComposite and get_substream_rstates

### PR DESCRIPTION
**Pull request by shabanian**
#4319
#4321

### CODE REVIEW

**Comment by shabanian** [comment]
@adri-romsor
@nouiz

File: theano/gof/op.py [hunk 1]
```diff
@@ -576,7 +576,11 @@ def _get_test_value(cls, v):
                 raise
             return ret
 
-        raise AttributeError('%s has no test value' % v)
+        sio = StringIO()
+        for subtr in getattr(v.tag, "trace", []):
+            traceback.print_list(subtr, sio)
+        detailed_err_msg = sio.getvalue()
+        raise AttributeError('%s has no test value %s' % (v, detailed_err_msg))
```
Comment (nouiz) [inline theano/gof/op.py]: Add a new line here for better display:

'%s has no test value %s' -> '%s has no test value \n%s'

File: theano/gof/op.py [hunk 2]
```diff
@@ -630,9 +634,14 @@ def __call__(self, *inputs, **kwargs):
...
                     elif config.compute_test_value == 'raise':
+                        sio = StringIO()
+                        for subtr in getattr(ins.tag, "trace", []):
+                            traceback.print_list(subtr, sio)
+                        detailed_err_msg = sio.getvalue()
+
                         raise ValueError(
-                            'Cannot compute test value: input %i (%s) of Op %s missing default value' %
-                            (i, ins, node))
+                            'Cannot compute test value: input %i (%s) of Op %s missing default value. %s' %
```
Comment (nouiz) [inline theano/gof/op.py]: idem.
In fact add somethine line:

default value. %s'  -> default value. \nStacktrace when the variable was created: %s'

File: theano/gof/utils.py [hunk 3]
```diff
@@ -99,6 +99,7 @@ def add_tag_trace(thing, user_line=None):
              "theano/scan_module/", "theano\\scan_module\\",
              "theano/sparse/", "theano\\sparse\\",
              "theano/typed_list/", "theano\\typed_list\\"]
+    skips = []
```
Comment (nouiz) [inline theano/gof/utils.py]: remove that, it was only for debugging.

File: theano/tensor/nnet/tests/test_bn.py [hunk 4]
```diff
@@ -25,6 +66,12 @@ def bn_ref(x, G, B, M, V):
     m = theano.tensor.vector('m')
     v = theano.tensor.vector('v')
 
+    x.tag.test_value = numpy.random.rand(2, 2)
```
Comment (nouiz) [inline theano/tensor/nnet/tests/test_bn.py]: remove those changes.

File: theano/gof/op.py (hunk 1 above)
Comment (nouiz) [inline theano/gof/op.py]: Travis have an error that I'm fixing. I'll make a PR to your PR with the fix.

Comment (nouiz) [inline theano/gof/op.py]: Here is the PR to your fix.

https://github.com/shabanian/Theano/pull/1

**Comment by nouiz** [comment]
travis fail with this error:

/home/travis/build/Theano/Theano/theano/gof/utils.py:104:1: W293 blank line contains whitespace

File: theano/tensor/nnet/tests/test_bn.py [hunk 5]
```diff
@@ -6,6 +6,47 @@
...
+        M = 1 + numpy.random.random([20]).astype('float32')
+        V = 1 + numpy.random.random([20]).astype('float32')
+
+        x = theano.tensor.matrix('x')
+        b = theano.tensor.vector('b')
+        g = theano.tensor.vector('g')
+        m = theano.tensor.vector('m')
+        v = theano.tensor.vector('v')
+
+        x.tag.test_value = numpy.random.rand(2, 2)
```
Comment (nouiz) [inline theano/tensor/nnet/tests/test_bn.py]: Cast the value to floatX like value.astype(theano.config.floatX).

Without this, the tests fail when floatX=float32.

File: theano/gof/op.py [hunk 6]
```diff
@@ -622,7 +622,7 @@ def __call__(self, *inputs, **kwargs):
                         detailed_err_msg = utils.get_variable_trace_string(ins)
 
                         raise ValueError(
-                            'Cannot compute test value: input %i (%s) of Op %s missing default value. %s' %
+                            'Cannot compute test value: input %i (%s) of Op %s missing default value. \nStacktrace when the variable was created: %s' %
```
Comment (nouiz) [inline theano/gof/op.py]: Remove this change. That line is added by the new call get_variable_trace_string.

**Pull request merged**